### PR TITLE
Flexible consumption V2 - Add components

### PR DESCRIPTION
**Pull request by Whenrow**
Increase the scope of flexible consumption, allowing to update confirmed MO's and consume components which weren't part of the initial BOM

https://www.odoo.com/web#id=2082448&action=333&active_id=966&model=project.task&view_type=form&menu_id=4720

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mrp/models/mrp_workorder.py [hunk 1]
```diff
@@ -401,6 +401,36 @@ def _defaults_from_finished_workorder_line(self, reference_lot_lines):
     def record_production(self):
         if not self:
             return True
+        # add missing move for extra component/byproduct
+        for line in self._workorder_line_ids():
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_workorder.py]: line 119 of the produce wizard, we handle exactly thise usecase: create a move for a owkorder without one

i suggest to remove the implementation in workorder you add here and in the produce wizard already present and add a shared one in abstract workorder

File: addons/mrp/models/stock_move.py [hunk 2]
```diff
@@ -17,6 +17,22 @@ class StockMoveLine(models.Model):
...
+        res = super(StockMoveLine, self).create(values)
+        for line in res:
+            # If the line is added in a done production, we need to map it
+            # manually to the produced move lines in order to see them in the
+            # traceability report
+            if line.move_id.raw_material_production_id and line.state == 'done':
+                mo = line.move_id.raw_material_production_id
+                if line.lot_produced_ids:
+                    produced_move_lines = mo.move_finished_ids.move_line_ids.filtered(lambda sml: sml.lot_id in line.lot_produced_ids)
+                    line.produce_line_ids = [(6, 0, [x for x in produced_move_lines.ids])]
```
Comment (sle-odoo) [inline addons/mrp/models/stock_move.py]: [(6, 0, produced_move_lines.ids)]

File: addons/mrp/views/mrp_workorder_views.xml [hunk 3]
```diff
@@ -168,9 +169,9 @@
                             <field name="move_id" invisible="1"/>
```
Comment (sle-odoo) [inline addons/mrp/views/mrp_workorder_views.xml]: if you have some workorder lines and you're in strict consumption, you can still add a line, can you check an error is properly raised?

File: addons/mrp/views/mrp_workorder_views.xml [hunk 4]
```diff
@@ -168,9 +169,9 @@
...
                         </tree>
                     </field>
-                    <h4 attrs="{'invisible': [('finished_workorder_line_ids', '=', [])]}">Finished Products</h4>
-                    <field name="finished_workorder_line_ids" attrs="{'invisible': [('finished_workorder_line_ids', '=', [])], 'readonly': [('state', 'in', 'cancel', 'done')]}">
-                        <tree editable="bottom" create="0" delete="0">
+                    <h4 groups="mrp.group_mrp_byproducts">Finished Products</h4>
```
Comment (sle-odoo) [inline addons/mrp/views/mrp_workorder_views.xml]: nope, if you don't enable the byproducts and you want to change some finished lots you will be stuck

File: addons/mrp/models/mrp_workorder.py [hunk 5]
```diff
@@ -663,6 +671,17 @@ class MrpWorkorderLine(models.Model):
     finished_workorder_id = fields.Many2one('mrp.workorder', 'Finished Product for Workorder',
         ondelete='cascade')
 
+    @api.model_create_multi
+    def create(self, values):
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_workorder.py]: vals_list

File: addons/mrp/models/mrp_workorder.py [hunk 6]
```diff
@@ -663,6 +671,17 @@ class MrpWorkorderLine(models.Model):
...
+        for line_data in values:
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_workorder.py]: for vals in vals_list

File: addons/mrp/models/mrp_workorder.py [hunk 7]
```diff
@@ -663,6 +671,17 @@ class MrpWorkorderLine(models.Model):
...
+            wo = self.env['mrp.workorder'].browse(line_data.get('raw_workorder_id'))
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_workorder.py]: raw_wokorder_id is not required, maybe you should let the create happen the check/raise

File: addons/mrp/views/mrp_workorder_views.xml [hunk 8]
```diff
@@ -168,9 +170,9 @@
                             <field name="move_id" invisible="1"/>
                         </tree>
                     </field>
-                    <h4 attrs="{'invisible': [('finished_workorder_line_ids', '=', [])]}">Finished Products</h4>
-                    <field name="finished_workorder_line_ids" attrs="{'invisible': [('finished_workorder_line_ids', '=', [])], 'readonly': [('state', 'in', 'cancel', 'done')]}">
-                        <tree editable="bottom" create="0" delete="0">
+                    <h4>Finished Products</h4>
+                    <field name="finished_workorder_line_ids" attrs="{'readonly': ['|', ('byproduct_group', '=', False), ('state', 'in', 'cancel', 'done')]}">
```
Comment (sle-odoo) [inline addons/mrp/views/mrp_workorder_views.xml]: add a compute is_finished_lines_editable?

File: addons/mrp/models/mrp_workorder.py [hunk 9]
```diff
@@ -401,6 +406,9 @@ def _defaults_from_finished_workorder_line(self, reference_lot_lines):
     def record_production(self):
         if not self:
             return True
+        # add missing move for extra component/byproduct
+        for line in self._workorder_line_ids():
+            line._set_move_id()
```
Comment (amoyaux) [inline addons/mrp/models/mrp_workorder.py]: You could try to move _set_move_id in abst_update_move_lines

File: addons/mrp/models/mrp_workorder.py [hunk 10]
```diff
@@ -164,6 +165,13 @@ def _set_dates_planned(self):
             'date_to': date_to,
         })
 
+    def _compute_is_finished_lines_editable(self):
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_workorder.py]: missing depends?

**Comment by amoyaux** [comment]
robodoo r+ rebase-ff

**Pull request closed without merging**